### PR TITLE
Start tracking code coverage progress, add function skipping annotations

### PR DESCRIPTION
**Pull request by rpatel-figure**
## Context
Now that optional record support has proven to be working, we can update the contracts to properly use `SkipIfRecordExists` for functions which are meant to initialize but never subsequently update a particular record. With that change, the basic functionality of all contracts should be available for use. Thus, before we add the remaining tests, let's start tracking code coverage to see the remaining progression and ensure development becomes fully test-driven.
### Major
- Modify constructors of eNote contracts to
  - use optional or required existing eNote record depending on whether it is allowed to update the record or not
  - use `@SkipIfRecordExists` in eNote record initialization contracts to ensure only eNote-updating contracts can do so
- Add [Kover](https://github.com/Kotlin/kotlinx-kover) to build workflow
### Minor
- DRY some data validation functions
- Improve the usage and/or signatures of some data validation functions
- Outline basic plan for remaining tests
- Add configurable code coverage & linting usage to build workflow dispatch
- Try increasing Kotest thread count from 2 to 4 & introducing environment variable for toggling edge case thoroughness
- Add concurrency for build workflow
- Update comments
- Improve some minor details in utility package code

### CODE REVIEW

File: contract/src/test/kotlin/io/provenance/scope/loan/contracts/RecordLoanContractUnitTest.kt [hunk 1]
```diff
@@ -387,18 +387,6 @@ class RecordLoanContractUnitTest : WordSpec({
             }
         }
     }
-    "recordServicingRights" When {
```
Comment (rpatel-figure) [inline contract/src/test/kotlin/io/provenance/scope/loan/contracts/RecordLoanContractUnitTest.kt:390]: To be tested in `OverwriteServicingRightsUnitTest`, since we never merge the input with the existing record.

File: contract/src/test/kotlin/io/provenance/scope/loan/contracts/UpdateENoteContractUnitTest.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            "throw an appropriate exception" {
+                // TODO: Implement
+            }
+        }
+        "given an invalid input" xshould {
+            "throw an appropriate exception" {
+                // TODO: Implement
+            }
+        }
+        "given an input which attempts to silently change properties of the existing eNote" xshould {
```
Comment (rpatel-figure) [inline contract/src/test/kotlin/io/provenance/scope/loan/contracts/UpdateENoteContractUnitTest.kt:17]: "silently change" == "provide the same checksum but alter other fields", see `documentModificationValidation`.

File: contract/src/test/kotlin/io/provenance/scope/loan/test/KotestHelpers.kt [hunk 3]
```diff
@@ -124,10 +165,21 @@ private val anyTimestampComponents: Arb<Pair<Long, Int>> = Arb.pair(
...
+            seconds == Timestamps.MIN_VALUE.seconds && nanoSeconds == Timestamps.MIN_VALUE.nanos
+        }
+    } else {
+        Arb.pair(
+            Arb.long(min = Timestamps.MIN_VALUE.seconds, max = now.epochSecond),
+            Arb.int(min = Timestamps.MIN_VALUE.nanos + 1, max = now.nano), // Minor sacrifice of case where nanos = 0, in exchange for faster runs
+        )
+    }.filterNot { (seconds, nanoSeconds) ->
+        seconds == 0L && nanoSeconds == 0
+    }
```
Comment (rpatel-figure) [inline contract/src/test/kotlin/io/provenance/scope/loan/test/KotestHelpers.kt:182]: I'm sure this seems a bit strange at a glance, and I do think it is 😅 but usage of `filterNot` definitely can have a significant effect on runtime, especially in the case of this function where it feeds into the tedious `anyValidLoanState` logic. The idea of the `runTestsExtended` is to allow the creation of better Arbs without forcing polynomial runtime. I have yet to apply it everywhere though because sometimes it becomes a lengthy overkill for basic things (like the UUID Arbs).

**Review by mnguyen-figure — DISMISSED** [review]

File: .github/workflows/publish-test-results.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
+name: Publish Test Results
```
Comment (cworsnop-figure) [inline .github/workflows/publish-test-results.yml:1]: :dank:

File: contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordENoteContract.kt [hunk 5]
```diff
@@ -16,11 +17,13 @@ import tech.figure.servicing.v1beta1.LoanStateOuterClass.ServicingData
 @Participants(roles = [PartyType.OWNER])
 @ScopeSpecification(["tech.figure.loan"])
 open class RecordENoteContract(
+    @Record(name = LoanScopeFacts.eNote, optional = true) val existingENote: ENote?,
```
Comment (cworsnop-figure) [inline contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordENoteContract.kt:20]: Im a bit confused why we'd have an existing enote and need to call recordEnoteContract? It seems we've made this contract for recording enotes OR servicing data? Should we not just have a contract for that instead?

File: contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordLoanContract.kt [hunk 6]
```diff
@@ -102,6 +103,7 @@ open class RecordLoanContract(
 
     @Function(invokedBy = PartyType.OWNER)
     @Record(LoanScopeFacts.servicingRights)
+    @SkipIfRecordExists(LoanScopeFacts.servicingRights)
```
[the remark this answers is not in the document]
Comment (cworsnop-figure) [reply contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordLoanContract.kt:106]: Since RecordEnoteContract doesn't mark this with `SkipIfRecordExists` doesnt that mean you could technically update it from there too?

File: contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordENoteContract.kt (hunk 5 above)
Comment (rpatel-figure) [reply contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordENoteContract.kt:20]: I think the original desire was for the `recordServicingData` function to be added to this contract as a convenience, but @dkneisly-figure or @mnguyen-figure may be able to speak to that further. I simply added the `@SkipIfRecordExists` to `recordENote` to prevent this contract from being able to further update the eNote after initializing it on the scope, and said annotation requires the record to be an optional value in the contract's constructor.

**Review by cworsnop-figure — APPROVED** [review]

File: contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordLoanContract.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (rpatel-figure) [reply contract/src/main/kotlin/io/provenance/scope/loan/contracts/RecordLoanContract.kt:106]: `RecordENoteContract` is able to record servicing _data_, not servicing _rights_. But I _think_ you are still correct that we should add the corresponding `@SkipIfRecordExists`  to `RecordENoteContract.recordServicingData`, as any future desire to alter something in the servicing data record other than appending loan states would heuristically warrant a separate contract. [I added the annotation](https://github.com/provenance-io/loan-package-contracts/compare/99a3d207601a9f568c8dd353f4c5c273ddc0684f..a9f70b74ca9971155cb94c924750a28e921e3559) so we can be notified by users if either such a new contract is needed or if the annotation should be removed for the convenience of `RecordENoteContract`.
